logger/src/lib.rs

```diff
@@ -1,159 +1,139 @@
+//       https://github.com/hyperledger/iroha/issues/3500
+pub fn init_global(configuration: &Configuration, terminal_colors: bool) -> Result<LoggerHandle> {
+    try_set_logger()?;
 
-fn setup_logger(configuration: &Configuration) -> Result<Telemetries> {
     let layer = tracing_subscriber::fmt::layer()
-        .with_ansi(configuration.terminal_colors)
+        .with_ansi(terminal_colors)
         .with_test_writer();
 
-    if configuration.compact_mode {
-        add_bunyan(configuration, layer.compact())
-    } else {
-        add_bunyan(configuration, layer)
+    match configuration.format {
+        Format::Full => step2(configuration, layer),
+        Format::Compact => step2(configuration, layer.compact()),
+        Format::Pretty => step2(configuration, layer.pretty()),
+        Format::Json => step2(configuration, layer.json()),
     }
 }
 
-fn bunyan_writer_create(destination: PathBuf) -> Result<Arc<std::fs::File>> {
-    OpenOptions::new()
-        .create(true)
-        .append(true)
-        .open(destination)
-        .wrap_err("Failed to create or open bunyan logs file")
-        .map(Arc::new)
+/// Returns once lazily initialised global logger for testing purposes.
+///
+/// # Panics
+/// If [`init_global`] or [`disable_global`] were called first.
+#[allow(clippy::needless_update)] // complications with `tokio-console` feature
+pub fn test_logger() -> LoggerHandle {
+    static LOGGER: OnceLock<LoggerHandle> = OnceLock::new();
+
+    LOGGER
+        .get_or_init(|| {
+            // NOTE: if this config should be changed for some specific tests, consider
+            // isolating those tests into a separate process and controlling default logger config
+            // with ENV vars rather than by extending `test_logger` signature. This will both remain
+            // `test_logger` simple and also will emphasise isolation which is necessary anyway in
+            // case of singleton mocking (where the logger is the singleton).
+            let config = Configuration {
+                level: Level::DEBUG,
+                format: Format::Pretty,
+                ..ConfigurationProxy::default().build().unwrap()
+            };
+
+            init_global(&config, true).expect(
+                "`init_global()` or `disable_global()` should not be called before `test_logger()`",
+            )
+        })
+        .clone()
+}
+
+/// Disables the logger globally, so that subsequent calls to [`init_global`] will fail.
+///
+/// Disabling logger is required in order to generate flamegraphs and flamecharts.
+///
+/// # Errors
+/// If global logger was already initialised/disabled.
+pub fn disable_global() -> Result<()> {
+    try_set_logger()
 }
 
-fn add_bunyan<L>(configuration: &Configuration, layer: L) -> Result<Telemetries>
+fn step2<L>(configuration: &Configuration, layer: L) -> Result<LoggerHandle>
 where
     L: tracing_subscriber::Layer<Registry> + Debug + Send + Sync + 'static,
```

Comment:
Well, the `L` type is different for each match arm calling it. It would either need dynamic dispatch (is `Layer` object safe?) or a static dispatch via an enum, both of which would bring more problems than a separated generic function for `step2` imo

Reply:
I see, thanks.